scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/KBucketsWithSubnetLimits.scala

```diff
@@ -0,0 +1,106 @@
+package io.iohk.scalanet.discovery.ethereum.v4
+
+import cats._
+import cats.implicits._
+import io.iohk.scalanet.discovery.hash.Hash
+import io.iohk.scalanet.discovery.ethereum.Node
+import io.iohk.scalanet.kademlia.{KBuckets, TimeSet}
+import io.iohk.scalanet.peergroup.Addressable
+import io.iohk.scalanet.peergroup.InetAddressOps._
+import java.net.InetAddress
+
+case class KBucketsWithSubnetLimits[A: Addressable](
+    table: KBuckets[Hash],
+    limits: KBucketsWithSubnetLimits.SubnetLimits,
+    tableLevelCounts: Map[InetAddress, Int],
+    bucketLevelCounts: Map[Int, Map[InetAddress, Int]]
+) {
+  import DiscoveryNetwork.Peer
+
+  def contains(peer: Peer[A]): Boolean =
+    table.contains(peer.kademliaId)
+
+  def touch(peer: Peer[A]): KBucketsWithSubnetLimits[A] =
+    copy(table = table.touch(peer.kademliaId))
+
+  def add(peer: Peer[A]): KBucketsWithSubnetLimits[A] =
+    if (contains(peer)) this
+    else {
+      val ip = subnet(peer)
+      val idx = getBucket(peer)._1
+
+      val tlc = tableLevelCounts |+| Map(ip -> 1)
+      val blc = bucketLevelCounts |+| Map(idx -> Map(ip -> 1))
+
+      val overTheLimit =
+        limits.prefixLength > 0 && (
+          limits.forTable > 0 && tlc(ip) > limits.forTable ||
+            limits.forBucket > 0 && blc(idx)(ip) > limits.forBucket
+        )
+
+      if (overTheLimit) this
+      else {
+        copy(
+          table = table.add(peer.kademliaId),
+          tableLevelCounts = tlc,
+          bucketLevelCounts = blc
+        )
+      }
+    }
+
+  def remove(peer: Peer[A]): KBucketsWithSubnetLimits[A] =
+    if (!contains(peer)) this
+    else {
+      val ip = subnet(peer)
+      val idx = getBucket(peer)._1
+
+      val tlc = tableLevelCounts |+| Map(ip -> -1) match {
+        case counts if counts(ip) <= 0 => counts - ip
+        case counts => counts
+      }
+      val blc = bucketLevelCounts |+| Map(idx -> Map(ip -> -1)) match {
```

Comment:
maybe we could add some comments about each case or export this to separate funcion with some docs as this logic is little dense ?

Reply:
Done, split it out into several methods.